source/extensions/transport_sockets/tls/utility.cc

```diff
@@ -119,6 +119,39 @@ int32_t Utility::getDaysUntilExpiration(const X509* cert, TimeSource& time_sourc
   return 0;
 }
 
+absl::optional<std::string> Utility::getX509ExtensionValue(const X509& cert,
+                                                           absl::string_view extension_name) {
+  X509_EXTENSIONS* extensions(X509_get0_extensions(&cert));
+
+  if (extensions == nullptr) {
+    return absl::nullopt;
+  }
+
+  const size_t extension_count = sk_X509_EXTENSION_num(extensions);
+
+  for (size_t i = 0; i < extension_count; ++i) {
+    X509_EXTENSION* extension = sk_X509_EXTENSION_value(extensions, i);
+
+    ASN1_OBJECT* extension_object = X509_EXTENSION_get_object(extension);
+    const size_t size = OBJ_obj2txt(nullptr, 0, extension_object, 0);
+    std::vector<char> buffer;
+    buffer.resize(size + 1);
+    OBJ_obj2txt(buffer.data(), buffer.size(), extension_object, 0);
+
+    if (std::string(buffer.data()) == extension_name) {
+      ASN1_OCTET_STRING* octet_string = X509_EXTENSION_get_data(extension);
+      const unsigned char* octet_string_data = octet_string->data;
+      long xlen;
+      int tag, xclass;
+      ASN1_get_object(&octet_string_data, &xlen, &tag, &xclass, octet_string->length);
+
+      return std::string(reinterpret_cast<const char*>(octet_string_data));
```

Comment:
will `octet_string_data` contains NUL? if so use length as well.

Reply:
using the returned length seems to be the safer option